=== modified file 'Makefile'
--- Makefile	2013-02-25 10:50:04 +0000
+++ Makefile	2013-11-22 09:00:36 +0000
@@ -39,14 +39,14 @@
 
 install-fedora:
 	mkdir -p "$(DESTDIR)/$(DRACUT_GROWROOT_D)"
-	for f in growroot.sh module-setup.sh ; do \
+	for f in growroot-dummy.sh growroot.sh module-setup.sh ; do \
 		install "growroot/$(DRACUT_GROWROOT_D)/$$f" \
 			"$(DESTDIR)/$(DRACUT_GROWROOT_D)/" ; \
 	done
 
 install-epel:
 	mkdir -p "$(DESTDIR)/$(DRACUT_GROWROOT_D)"
-	for f in growroot.sh install ; do \
+	for f in growroot-dummy.sh growroot.sh install ; do \
 		install "growroot/$(DRACUT_GROWROOT_D)/$$f" \
 			"$(DESTDIR)/$(DRACUT_GROWROOT_D)/" ; \
 	done

=== added file 'growroot/dracut/modules.d/50growroot/growroot-dummy.sh'
--- growroot/dracut/modules.d/50growroot/growroot-dummy.sh	1970-01-01 00:00:00 +0000
+++ growroot/dracut/modules.d/50growroot/growroot-dummy.sh	2013-11-22 09:01:17 +0000
@@ -0,0 +1,3 @@
+#!/bin/sh
+# Dummy script, to make sure systemd executes the cmdline stage (which exports
+# the 'root' variable required by growroot)

=== modified file 'growroot/dracut/modules.d/50growroot/growroot.sh'
--- growroot/dracut/modules.d/50growroot/growroot.sh	2013-11-21 08:15:17 +0000
+++ growroot/dracut/modules.d/50growroot/growroot.sh	2013-11-22 11:57:24 +0000
@@ -1,9 +1,7 @@
 #!/bin/sh
 
 # Environment variables that this script relies upon:
-# - NEWROOT
-
-. /lib/dracut-lib.sh
+# - root
 
 _info() {
 	echo "growroot: $*"
@@ -19,35 +17,11 @@
 	exit 1
 }
 
-# This runs right before exec of /sbin/init, the real root is already mounted
-# at NEWROOT
-_growroot()
-{
-	local out rootdev rootmnt rootfs opts unused rootdisk partnum
-
-	# If a file indicates we should do nothing, then just return
-	for file in /var/lib/cloud/instance/root-grown /etc/growroot-disabled \
-		/etc/growroot-grown ; do
-		if [ -f "${NEWROOT}${file}" ] ; then
-			_info "${file} exists, nothing to do"
-			return
-		fi
-	done
-
-	# Get the root device, root filesystem and mount options
-	if ! out=$(awk '$2 == mt { print }' "mt=${NEWROOT}" < /proc/mounts) ; then
-		_warning "${out}"
-		return
-	fi
-
-	# Need to do it this way, can't use '<<< "${out}"' since RHEL6 doesn't
-	# seem to understand it
-	read rootdev rootmnt rootfs opts unused <<EOF
-${out}
-EOF
-	if [ -z "${rootdev}" -o -z "${rootmnt}" -o -z "${rootfs}" -o \
-		-z "${opts}" ] ; then
-		_warning "${out}"
+_growroot() {
+	# Remove 'block:' prefix and find the root device
+	rootdev=$(readlink -f "${root#block:}")
+	if [ -z "${rootdev}" ] ; then
+		_warning "unable to find root device"
 		return
 	fi
 
@@ -79,24 +53,19 @@
 		rootdisk=${rootdisk%p}
 	fi
 
-	# Do a growpart dry run and exit if it fails or doesn't have anything to do
+	# Do a growpart dry run and exit if it fails or doesn't have anything
+	# to do
 	if ! out=$(growpart --dry-run "${rootdisk}" "${partnum}") ; then
 		_info "${out}"
 		return
 	fi
 
-	# There's something to do so unmount and re-partition
-	if ! umount "${NEWROOT}" ; then
-		_warning "Failed to umount ${NEWROOT}"
-		return
-	fi
-
 	# Wait for any of the initial udev events to finish otherwise growpart
 	# might fail
-	udevsettle
+	udevadm settle --timeout=30
 
 	# Resize the root partition
-	if out=$(growpart "${rootdisk}" "${partnum}" 2>&1) ; then
+	if out=$(growpart --update off "${rootdisk}" "${partnum}" 2>&1) ; then
 		_info "${out}"
 	else
 		_warning "${out}"
@@ -104,17 +73,15 @@
 	fi
 
 	# Wait for the partition re-read events to complete so that the root
-	# partition is available for remounting
-	udevsettle
-
-	# Remount the root filesystem
-	mount -t "${rootfs}" -o "${opts}" "${rootdev}" "${NEWROOT}" || \
-		_fatal "Failed to re-mount ${rootdev}, this is bad"
-
-	# Write to /etc/growroot-grown, most likely this wont work (read-only)
-	{
-		date --utc > "${NEWROOT}/etc/growroot-grown"
-	} >/dev/null 2>&1
+	# partition is available for mounting
+	udevadm settle --timeout=30
+
+	# Add the root partition if it didn't come back on its own
+	if ! [ -e "${rootdev}" ] ; then
+		partx --add --nr "${partnum}" "${rootdisk}" || \
+			_warning "failed to add root device ${rootdev}"
+		udevadm settle --timeout=30
+	fi
 }
 
 _growroot

=== modified file 'growroot/dracut/modules.d/50growroot/install'
--- growroot/dracut/modules.d/50growroot/install	2013-02-25 10:48:00 +0000
+++ growroot/dracut/modules.d/50growroot/install	2013-11-22 09:01:19 +0000
@@ -1,7 +1,11 @@
 #!/bin/bash
-
-inst_hook pre-pivot 00 "$moddir/growroot.sh"
-
+# Dracut install script for RHEL6
+ 
+inst_hook cmdline 99 "$moddir/growroot-dummy.sh"
+inst_hook pre-mount 99 "$moddir/growroot.sh"
+ 
 dracut_install awk
 dracut_install growpart
+dracut_install partx
+dracut_install readlink
 dracut_install sfdisk

=== modified file 'growroot/dracut/modules.d/50growroot/module-setup.sh'
--- growroot/dracut/modules.d/50growroot/module-setup.sh	2013-02-27 01:45:33 +0000
+++ growroot/dracut/modules.d/50growroot/module-setup.sh	2013-11-22 09:01:19 +0000
@@ -1,4 +1,5 @@
 #!/bin/bash
+# Dracut install script for Fedora
 
 check() {
 	return 0
@@ -9,10 +10,13 @@
 }
 
 install() {
-	inst_hook pre-pivot 00 "$moddir/growroot.sh"
+	inst_hook cmdline 99 "$moddir/growroot-dummy.sh"
+	inst_hook pre-mount 99 "$moddir/growroot.sh"
 
 	dracut_install awk
 	dracut_install growpart
+	dracut_install partx
+	dracut_install readlink
 	dracut_install sfdisk
 	dracut_install sgdisk
 }

